batch-processing/pipeline.go

```diff
@@ -0,0 +1,353 @@
+	ErrImageExistsInS3 = errors.New("image already exists in S3, skipping")
+)
+
+// Pipeline struct
+type Pipeline struct {
+	config     *Config
+	workers    map[string]int
+	channels   map[string]chan *Out
+	maxRetries uint64    // Max number of retries for downloading and uploading an image
+	urlMap     *sync.Map // this will record urls that are already processed so we don't download again.
+}
+
+type Task func(wg *sync.WaitGroup)
+
+func (p *Pipeline) Read(csvReader *csv.Reader) {
+	// Read the input CSV file
+	// For each line, send the URL to the download channel
+	defer close(p.channels[READ])
+
+	csvReader.FieldsPerRecord = 1
+
+	for {
+		row, err := csvReader.Read()
+
+		if err == io.EOF {
+			break
+		}
+
+		if err != nil {
+			p.channels[READ] <- &Out{Err: err}
+			continue
+		}
+		p.channels[READ] <- &Out{Url: row[0]}
+	}
+	log.Println("Read done")
+}
+
+func (p *Pipeline) Download(wg *sync.WaitGroup) {
+	// Download the image from the URL
+	// Send the image to the convert channel
+	defer wg.Done()
+	var (
+		row  *Out
+		ok   bool
+		body io.Reader
+		hash string
+		ext  string
+		err  error
+	)
+	for {
+
+		row, ok = <-p.channels[READ]
+		if !ok {
+			break
+		}
+		if row.Err != nil {
+			p.channels[DOWNLOAD] <- row
+			continue
+		}
+
+		// Check if the URL has already been processed
+		// If it has, skip it
+		_, ok := p.urlMap.Load(row.Url)
+		if ok {
+			p.channels[DOWNLOAD] <- &Out{
+				Url: row.Url,
+				Err: ErrDuplicateURL,
+			}
+			continue
+		} else {
+			p.urlMap.Store(row.Url, true)
+		}
+
+		body, ext, hash, err = DownloadWithBackoff(row.Url, p.maxRetries)
+
+		if err != nil {
+			row.Err = err
+			p.channels[DOWNLOAD] <- row
+			continue
+		}
+
+		key := fmt.Sprintf("%s-converted.%s", hash, ext)
+
+		// Check if the image hash is already in the S3 bucket
+		_, err = p.config.Aws.GetObject(&s3.GetObjectInput{
+			Bucket: &p.config.Aws.s3bucket,
+			Key:    &key,
+		})
+
+		if err == nil {
+			row.Err = ErrImageExistsInS3
+			row.S3url = fmt.Sprintf("https://%s.s3.amazonaws.com/%s", p.config.Aws.s3bucket, hash)
+			p.channels[DOWNLOAD] <- row
+			continue
+		}
+
+		//Create an empty file
+		inputPath := fmt.Sprintf("/outputs/%s.%s", hash, ext)
+		outputPath := fmt.Sprintf("/outputs/%s", key)
+
+		file, err := os.Create(inputPath)
+		if err != nil {
+			p.channels[DOWNLOAD] <- &Out{
+				Url: row.Url,
+				Err: err,
+			}
+			continue
+		}
+		defer file.Close()
+
+		//Write bytes to the file
+		_, err = io.Copy(file, body)
+		if err != nil {
+			p.channels[DOWNLOAD] <- &Out{
+				Url: row.Url,
+				Err: err,
+			}
+			continue
+		}
+
+		p.channels[DOWNLOAD] <- &Out{
+			Url:    row.Url,
+			Input:  inputPath,
+			Output: outputPath,
+		}
+	}
+}
+
+func (p *Pipeline) Convert(wg *sync.WaitGroup) {
+	// Convert the image
+	// Send the image to the upload channel
+	defer wg.Done()
+	for {
+		row, ok := <-p.channels[DOWNLOAD]
+		if !ok {
+			break
+		}
+		if row.Err != nil {
+			p.channels[CONVERT] <- &Out{
```

Comment:
Why are you re-constructing these errors:
```go
<- &Out{
    Url: row.Url,
    Err: row.Err,
    Input: row.Input,
    Output: "",
}
```

rather than just sending the same pointer on the new channel:
```go
<- row
```

Reply:
I did it because I was setting the path of `Output` (convert file path) from download stage, so I was essentially un-setting it here. 

This is no longer the case with refactoring. 

Thanks!